tf-dataflow/real_data_loader.py

```diff
@@ -0,0 +1,265 @@
+
+    if len(rgb_imgs) == 0:
+        LOG.error("No images found!")
+        return
+
+    rgb_image_batches = []
+    batch_size = int(len(rgb_imgs) / _concurrency)
+    for i in range(_concurrency):
+        rgb_image_batches.append(set(random.sample(rgb_imgs, batch_size)))
+        rgb_imgs -= rgb_image_batches[i]
+        LOG.info(
+            "Starting thread {} with {} images.".format(
+                i, len(rgb_image_batches[i])
+            )
+        )
+        _thread.start_new_thread(
+            _write_records_batch, (rgb_image_batches[i], annotations, i)
+        )
+
+    while 1:
+        pass
+
+
+def _write_records_batch(images, annotations, batch):
+    train = set(random.sample(images, int(len(images) * _train_val_ratio)))
+    val = images - train
+
+    training_record_name = "training_{:06d}"
+    training_record_name += "_batch_" + str(batch) + ".tfrecord"
+    _write_records(train, annotations, training_record_name, batch)
+
+    validation_record_name = "validation_{:06d}"
+    validation_record_name += "_batch_" + str(batch) + ".tfrecord"
+    _write_records(val, annotations, validation_record_name, batch)
+
+
+def _write_records(images, annotations, naming_convention, batch):
+    data_written_mb = 0.0
+    record_num = 0
+    writer = tf.io.TFRecordWriter(
+        os.path.join(_out_path, naming_convention.format(record_num))
+    )
+    for img in images:
+        example = convert_to_example(img, annotations[img])
+        writer.write(example)
+        data_written_mb += sys.getsizeof(example) / _bytes_in_mb
+        if data_written_mb > _record_size_max_mb:
+            writer.close()
+            record_num += 1
+            writer = tf.io.TFRecordWriter(
+                os.path.join(_out_path, naming_convention.format(record_num))
+            )
+            data_written_mb = 0
+    writer.close()
+
+
+def _bytes_feature(value):
+    """Returns a bytes_list from a string / byte."""
+    if isinstance(value, type(tf.constant(0))):
+        value = (
+            value.numpy()
+        )  # BytesList won't unpack a string from an EagerTensor.
+    return tf.train.Feature(bytes_list=tf.train.BytesList(value=[value]))
+
+
+def _float_feature(value):
+    return _float_list_feature([value])
+
+
+def _float_list_feature(value):
+    """Returns a float_list from a float / double."""
+    return tf.train.Feature(float_list=tf.train.FloatList(value=value))
+
+
+def _int64_feature(value):
+    """Returns an int64_list from a bool / enum / int / uint."""
+    return _int64_list_feature([value])
+
+
+def _int64_list_feature(value):
+    return tf.train.Feature(int64_list=tf.train.Int64List(value=value))
+
+
+def _convert_to_example(filename, image_buffer, height, width, labels, bboxes):
+    colorspace = b"RGB"
+    channels = 3
+    image_format = b"JPG"
+    xmins, xmaxes, ymins, ymaxes = bboxes
+
+    example = tf.train.Example(
+        features=tf.train.Features(
+            feature={
```

Comment:
DRY! This protobuf serialization definition is shared between UnityGroceries-Real and synthetic dataset. Can you make sure these code can be reused?

Reply:
This script is modified specifically for UnityGroceries-Real dataset and not identical with the method for synthetic dataset. Tested and verified the transformation functionality.